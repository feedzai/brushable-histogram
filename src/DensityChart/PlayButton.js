import React, { PureComponent } from "react";
import PropTypes from "prop-types";

// These icon paths were copied from antd
const PAUSE_ICON_PATH = `M512 64C264.6 64 64 264.6 64 512s200.6 448 448 448 448-200.6 448-448S759.4
64 512 64zm0 820c-205.4 0-372-166.6-372-372s166.6-372 372-372 372 166.6 372 372-166.6 372-372
372zm-88-532h-48c-4.4 0-8 3.6-8 8v304c0 4.4 3.6 8 8 8h48c4.4 0 8-3.6 8-8V360c0-4.4-3.6-8-8-8zm224
0h-48c-4.4 0-8 3.6-8 8v304c0 4.4 3.6 8 8 8h48c4.4 0 8-3.6 8-8V360c0-4.4-3.6-8-8-8z`;

const PLAY_ICON_PATHS = [
    `M512 64C264.6 64 64 264.6 64 512s200.6 448 448 448 448-200.6 448-448S759.4 64 512 64zm0 820c-205.4
    0-372-166.6-372-372s166.6-372 372-372 372 166.6 372 372-166.6 372-372 372z`,
    `M719.4 499.1l-296.1-215A15.9 15.9 0 0 0 398 297v430c0 13.1 14.8 20.5 25.3 12.9l296.1-215a15.9
    15.9 0 0 0 0-25.8zm-257.6 134V390.9L628.5 512 461.8 633.1z`
];

/**
 * PlayButton
 *
 * Renders a play button that allows to play a time-lapse of the events
 * in the histogram.
 *
 * @author Beatriz Malveiro Jorge (beatriz.jorge@feedzai.com)
 * @author Victor Fernandes (victor.fernandes@feedzai.com)
 * @author Luis Cardoso (luis.cardoso@feedzai.com)
 */

export default class PlayButton extends PureComponent {
    static propTypes = {
        width: PropTypes.number.isRequired,
        brushDomainMax: PropTypes.number.isRequired,
        brushDomainMin: PropTypes.number.isRequired,
        densityChartXScale: PropTypes.func.isRequired,
        moveBrush: PropTypes.func.isRequired,
        frameStep: PropTypes.number,
        frameDelay: PropTypes.number
    };

    static defaultProps = {
        frameStep: 0.025,
        frameDelay: 500
    };

    state = {
        play: false
    };

    /**
     * Handles click on play button. Defines start and end for
     * the domain-lapse and triggers _playLapse to play frames
     * at set intervals.
     *
     * @private
     */
    _onClickPlay = () => {
        const { width, densityChartXScale, brushDomainMax, brushDomainMin, frameStep } = this.props;
        const brushedMaxRange = densityChartXScale(brushDomainMax);
        const brushedMinRange = densityChartXScale(brushDomainMin);
        const frameStart = brushedMinRange;

        const playEnd = width;
        const playStep = width * frameStep;

        this.frameEnd = frameStart;

        if (brushedMaxRange === playEnd) {
            this.frameEnd = frameStart;
        } else {
            this.frameEnd = brushedMaxRange;
        }

        this.setState({
            play: true
        }, () => this._playLapseAtInterval(frameStart, playEnd, playStep));
    };

    /**
     * Handles click on stop button. Will clear interval
     * of funtion playInterval playing domain-lapse frames
     *
     * @private
     */
    _onClickStop = () => {
        this._stopLapse();
    };

    /**
     * Plays a frame of the domain-lapse. Updates subset of domain
     * to be displayed and moves brush to new domain.
     *
     * @param {Number} start
     * @param {Number} end
     * @param {Number} step
     * @private
     */
    _playFrame(start, end, step) {
        // If end of frame is at end of play region then stop domain-lapse

        if (this.frameEnd >= end) {
            this._stopLapse();
            return;
        }

        // Check if adding a step will surprass max domain.
        if (this.frameEnd + step >= end) {
            // If so, set max frame to be max domain
            this.frameEnd = end;
        } else {
            // Otherwise just add a step
            this.frameEnd += step;
        }
        const domain = [start, this.frameEnd];

        // Move brush to new frame location
        this.props.moveBrush(domain);
    }

    /**
     * Plays domain-lapse by calling _playFrame at interval
     * to play each frame.
     *
     * @param {Number} start
     * @param {Number} end
     * @param {Number} step
     * @private
     */
    _playLapseAtInterval(start, end, step) {
        this.playInterval = setInterval(() => {
            this._playFrame(start, end, step);
        }, this.props.frameDelay);
    }

    /**
     *  Stops domain-lapse at current frame. This
     * is done by clearing the timeInterval in this.playInterval.
     *
     * @private
     */
    _stopLapse() {
        this.setState({
            play: false
        }, () => clearInterval(this.playInterval));
    }

    render() {
        const onClick = this.state.play ? this._onClickStop : this._onClickPlay;

        let iconElement;

        if (this.state.play) {
            iconElement = (
                <svg viewBox="64 64 896 896" width="1em" height="1em" fill="currentColor">
                    <path d={PAUSE_ICON_PATH}/>
                </svg>
            );
        } else {
            iconElement = (
                <svg viewBox="64 64 896 896" width="1em" height="1em" fill="currentColor">
                    <path d={PLAY_ICON_PATHS[0]}/>
                    <path d={PLAY_ICON_PATHS[1]}/>
                </svg>
            );
        }

        return (<>
            <button type="button" className="ant-btn fdz-css-play-btn ant-btn-icon-only" onClick={onClick}>
                <i className="anticon anticon-play-circle">
                    {iconElement}
                </i>
            </button>
        </>);
    }
}
