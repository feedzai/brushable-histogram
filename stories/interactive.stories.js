/**
 * 2019, Feedzai
 *
 * This source code is licensed under the LGPL license found in the
 * LICENSE file in the root directory of this source tree.
 */

import React from "react";
import { storiesOf } from "@storybook/react";
import { withKnobs, object } from "@storybook/addon-knobs";
import Histogram from "../src/index";
import "../src/Histogram/Histogram.scss";

/**
 * @author Victor Fernandes (victor.fernandes@feedzai.com)
 * @since 1.7.0
 */

const stories = storiesOf("Histogram - Interactive data", module);

stories.addDecorator(withKnobs);

stories
    .add("Basic example", () => (
        <Histogram
            data={object("Data", [
                { timestamp: 1170070000000, total: 100 },
                { timestamp: 1270070000000, total: 23 },
                { timestamp: 1370070000000, total: 100 },
                { timestamp: 1470070000000, total: 23 },
                { timestamp: 1570070000000, total: 100 },
                { timestamp: 1670070000000, total: 23 },
                { timestamp: 1770070000000, total: 400 },
                { timestamp: 1870070000000, total: 200 }
            ])}
            xAccessor={(datapoint) => datapoint.timestamp}
            yAccessor={(datapoint) => datapoint.total}
            onIntervalChange={(domain) => console.log(domain)}
        />
    ));
