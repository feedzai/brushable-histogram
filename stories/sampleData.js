const NUMBER_OF_POINTS = 60000;
const startTimestamp = 1533164400000;

/**
 * Calculates some sample data
 *
 * @param {number} numberOfPoints
 * @returns {Array.<Object>}
 */
export function calculate(numberOfPoints) {
    const data = [];

    for (let i = 0; i < numberOfPoints; i++) {
        data.push({
            "timestamp": startTimestamp + Math.abs(Math.floor(Math.cos(i) * 1000000000)),
            "total": Math.abs(Math.sin(i))
        });
    }

    return data;
}

export default calculate(NUMBER_OF_POINTS);

export const smallSample = calculate(10);
