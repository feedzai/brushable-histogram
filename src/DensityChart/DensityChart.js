import React, { PureComponent } from "react";
import PropTypes from "prop-types";
import PlayButton from "./PlayButton";
import { event as d3Event, select as d3Select } from "d3-selection";
import {
    clearCanvas,
    drawRect,
    getRenderContext
} from "../canvasRenderUtils";
import {
    havePropsChanged
} from "../utils";
import { brushX } from "d3-brush";

/**
 * DensityChart
 *
 * Plots a density strip plot for context when brushing and zooming the histogram.
 *
 * @author Beatriz Malveiro Jorge (beatriz.jorge@feedzai.com)
 * @author Victor Fernandes (victor.fernandes@feedzai.com)
 * @author Luis Cardoso (luis.cardoso@feedzai.com)
 */

export default class DensityChart extends PureComponent {
    static propTypes = {
        data: PropTypes.arrayOf(PropTypes.object).isRequired,
        width: PropTypes.number.isRequired,
        height: PropTypes.number.isRequired,
        padding: PropTypes.number.isRequired,
        brushDomainMax: PropTypes.number.isRequired,
        brushDomainMin: PropTypes.number.isRequired,
        densityChartXScale: PropTypes.func.isRequired,
        onDomainChanged: PropTypes.func.isRequired,
        xAccessor: PropTypes.func.isRequired,
        frameStep: PropTypes.number,
        frameDelay: PropTypes.number,
        brushDensityChartColor: PropTypes.string,
        brushDensityChartFadedColor: PropTypes.string,
        renderPlayButton: PropTypes.bool
    };

    static defaultProps = {
        renderPlayButton: true,
        brushDensityChartColor: "rgba(33, 150, 243, 0.2)",
        brushDensityChartFadedColor: "rgba(176, 190, 197, 0.2)"
    };

    constructor(props) {
        super(props);

        this.densityChartRef = React.createRef();
        this.densityBrushRef = React.createRef();
    }

    componentDidMount() {
        this.densityChartCanvasContext = getRenderContext(this.densityChartRef.current);

        const { width, height, densityChartXScale } = this.props;

        this.brush = brushX()
            .extent([
                [0, 0],
                [width, height]
            ])
            .on("brush end", this._onResizeBrush);

        this._updateBrush();

        this._moveBrush(densityChartXScale.range());

        this._drawDensityChart();
    }

    componentDidUpdate(prevProps) {
        const { brushDomainMin, brushDomainMax, densityChartXScale } = this.props;

        this._updateBrush();

        this._moveBrush([densityChartXScale(brushDomainMin), densityChartXScale(brushDomainMax)]);

        // We only need to re-render the density chart if the data, the weight, the height or
        // the chart x scale have changed.
        if (this._shouldRedrawDensityChart(prevProps)) {
            this._drawDensityChart();
        }
    }

    componentWillUnmount() {
        clearInterval(this.playInterval);
        this.brush.on("brush end", null); // This is the way to unbind events in d3
    }

    /**
     * Handles brush events. It will update this.state.brushedDomain according to the
     * transformation on the event.
     *
     * @private
     */
    _onResizeBrush = () => {
        // This occurs always when the user change the brush domain manually

        const event = this._getD3Event();

        if (event.sourceEvent && event.sourceEvent.type === "zoom") {
            return;
        }

        let brushSelection;

        if (Array.isArray(event.selection)) {
            brushSelection = event.selection;
        } else {
            // When we don't have any selection we should select everything
            brushSelection = this.props.densityChartXScale.range();
        }

        this.props.onDomainChanged(brushSelection);
    };

    /**
     * Returns the D3 event object
     *
     * Used for stubbing in tests.
     *
     * @returns {Object|null}
     */
    _getD3Event() {
        return d3Event;
    }

    /**
     * Reapplies the brush
     * @private
     */
    _updateBrush() {
        if (this.props.data.length === 0) {
            return;
        }

        d3Select(this.densityBrushRef.current)
            .call(this.brush);
    }

    /**
     * Moves brush on density strip plot to given domain
     * @private
     * @param {Array<Number>} domain
     */
    _moveBrush = (domain) => {
        if (this.props.data.length === 0) {
            return;
        }

        d3Select(this.densityBrushRef.current)
            .call(this.brush.move, domain);
    };

    /**
     * Returns whenever it is necessary to re-render the density chart, based on the current and previous
     * props.
     * @param {Object} prevProps
     * @returns {boolean}
     */
    _shouldRedrawDensityChart(prevProps) {
        return havePropsChanged(this.props, prevProps, [
            "brushDomainMin",
            "brushDomainMax",
            "data",
            "width",
            "height",
            "densityChartXScale"
        ]);
    }

    /**
     * Draws density strip plot in canvas.
     * (Using canvas instead of svg for performance reasons as number of datapoints
     * can be very large)
     *
     * @private
     */
    _drawDensityChart() {
        const {
            width,
            height,
            densityChartXScale,
            brushDomainMax,
            brushDomainMin,
            xAccessor,
            data,
            brushDensityChartColor,
            brushDensityChartFadedColor
        } = this.props;

        clearCanvas(this.densityChartCanvasContext, width, height);

        for (let i = 0; i < data.length; ++i) {
            const x = xAccessor(data[i]);
            const isInsideOfBrushDomain = x >= brushDomainMin && x < brushDomainMax;

            drawRect(
                this.densityChartCanvasContext, // canvas context
                densityChartXScale(x), // x
                0, // y
                2, // width
                height, // height
                {
                    fillStyle: isInsideOfBrushDomain ? brushDensityChartColor : brushDensityChartFadedColor
                }
            );
        }
    }

    /**
     * Renders the play button that allows to replay a time-lapse of the events.
     * @returns {React.Element|null}
     */
    _renderPlayButton() {
        const { width, densityChartXScale, brushDomainMax, brushDomainMin,
            frameStep, frameDelay, renderPlayButton } = this.props;

        if (!renderPlayButton) {
            return null;
        }

        return (
            <PlayButton
                width={width}
                densityChartXScale={densityChartXScale}
                brushDomainMax={brushDomainMax}
                brushDomainMin={brushDomainMin}
                frameStep={frameStep}
                frameDelay={frameDelay}
                moveBrush={this._moveBrush}
            />
        );
    }

    render() {
        let leftPadding = 0;

        const { width, height, padding } = this.props;

        if (!this.props.renderPlayButton) {
            leftPadding = padding * 2;
        }

        const densityChartCanvasStyle = { left: leftPadding };

        return (
            <div className="fdz-css-graph-histogram-density__wrapper" >
                {this._renderPlayButton()}
                <div className="fdz-css-graph-histogram-density" style={{ position: "relative" }}>
                    <canvas
                        ref={this.densityChartRef}
                        className="fdz-css-graph-histogram-density__canvas"
                        width={width}
                        height={height}
                        style={densityChartCanvasStyle}
                        aria-label="Density Chart"
                    />
                    <svg
                        ref={this.densityBrushRef}
                        className="fdz-css-graph-histogram-brush"
                        width={width}
                        height={height}
                        style={{ position: "absolute", left: leftPadding, top: 0 }}
                        alt="Density Chart Brush"
                    />
                </div>
            </div>
        );
    }
}
