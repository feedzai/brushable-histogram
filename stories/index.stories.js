import React, { Fragment } from "react";
import { timeFormat } from "d3-time-format";

import { storiesOf } from "@storybook/react";
import { withKnobs, number, object } from "@storybook/addon-knobs";

import sampleData from "./sampleData";
import Histogram from "../src/index";
import "../src/Histogram/Histogram.scss";

const stories = storiesOf("Histogram", module);

stories.addDecorator(withKnobs);

const formatDayFull = timeFormat("%e %b'%y");

function formatContextInterval(dateStart, dateEnd) {
    const formatStart = formatDayFull(dateStart);
    const formatEnd = formatDayFull(dateEnd);

    // If the formatted start and end date are the same then just display one
    if (formatStart === formatEnd) {
        return formatStart;
    }

    // Otherwise display the interval limits.
    return `${formatStart} - ${formatEnd}`;
}

function histogramTooltipBar(bar) {
    return (
        <Fragment>
            <div className="fdz-css-graph-histogram-bars--tooltip-value">
                {Math.floor(bar.yValue)} Events
            </div>
            <div className="fdz-css-graph-histogram-bars--tooltip-dates">
                {formatContextInterval(bar.x0, bar.x1)}
            </div>
        </Fragment>
    );
}

stories
    .add("Basic example", () => (
        <Histogram
            data={sampleData}
            xAccessor={(datapoint) => datapoint.timestamp}
            yAccessor={(datapoint) => datapoint.total}
        />
    ))
    .add("With a custom tooltip", () => (
        <Histogram
            data={sampleData}
            xAccessor={(datapoint) => datapoint.timestamp}
            yAccessor={(datapoint) => datapoint.total}
            tooltipBarCustomization={histogramTooltipBar}
        />
    ))
    .add("With a custom axis formatters", () => (
        <Histogram
            data={sampleData}
            xAccessor={(datapoint) => datapoint.timestamp}
            xAxisFormatter={formatDayFull}
            yAxisFormatter={(value) => (value > 0 ? `${value}$` : "")}
            yAccessor={(datapoint) => datapoint.total}
        />
    ))
    .add("Without the play button", () => (
        <Histogram
            data={sampleData}
            xAccessor={(datapoint) => datapoint.timestamp}
            yAccessor={(datapoint) => datapoint.total}
            renderPlayButton={false}
        />
    ))
    .add("With a custom a height", () => (
        <Histogram
            data={sampleData}
            xAccessor={(datapoint) => datapoint.timestamp}
            yAccessor={(datapoint) => datapoint.total}
            height={number("height (min 150)", 300)}
            tooltipBarCustomization={histogramTooltipBar}
        />
    ))
    .add("With no data", () => (
        <Histogram
            data={[]}
            xAccessor={(datapoint) => datapoint.timestamp}
            yAccessor={(datapoint) => datapoint.total}
            tooltipBarCustomization={histogramTooltipBar}
        />
    ))
    .add("Change data in order to update histogram", () => (
        <Histogram
            data={object("Data", [
                { timestamp: 1170070000000, total: 100 },
                { timestamp: 1270070000000, total: 23 },
                { timestamp: 1370070000000, total: 100 },
                { timestamp: 1470070000000, total: 23 },
                { timestamp: 1570070000000, total: 100 },
                { timestamp: 1670070000000, total: 23 },
                { timestamp: 1770070000000, total: 400 },
                { timestamp: 1870070000000, total: 200 }
            ])}
            xAccessor={(datapoint) => datapoint.timestamp}
            yAccessor={(datapoint) => datapoint.total}
            onIntervalChange={(domain) => console.log(domain)}
        />
    ));
