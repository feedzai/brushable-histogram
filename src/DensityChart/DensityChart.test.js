import React from "react";
import { mount } from "enzyme";
import { max as d3Max, min as d3Min } from "d3-array";
import { scaleTime } from "d3-scale";
import DensityChart from "./DensityChart";
import PlayButton from "./PlayButton";
import { smallSample } from "../../stories/sampleData";

jest.mock("../canvasRenderUtils", () => ({
    drawRect: () => {},
    clearCanvas: () => {},
    getRenderContext: () => ({})
}));

const d3Event = global.d3Event;

let onDomainChanged, xAccessor, width, brushDomainMin, brushDomainMax, densityChartXScale;

beforeEach(() => {
    onDomainChanged = jest.fn();
    xAccessor = (elm) => elm.timestamp;
    width = 1000;
    brushDomainMin = d3Min(smallSample, xAccessor);
    brushDomainMax = d3Max(smallSample, xAccessor);
    densityChartXScale = scaleTime()
        .domain([ brushDomainMin, brushDomainMax])
        .range([ 0, width ]);
});

describe("_onResizeBrush", () => {
    let instance;

    beforeEach(() => {
        const wrapper = mount(<DensityChart
            data={smallSample}
            width={width}
            height={50}
            padding={10}
            brushDomainMax={brushDomainMax}
            brushDomainMin={brushDomainMin}
            densityChartXScale={densityChartXScale}
            onDomainChanged={onDomainChanged}
            xAccessor={xAccessor}
        />);

        instance = wrapper.instance();
    });

    it("does nothing if the event type is zoom", () => {
        instance._getD3Event = () => ({
            sourceEvent: {
                type: "zoom"
            }
        });

        instance._onResizeBrush();

        expect(onDomainChanged.mock.calls.length).toBe(2);
    });

    it("calls props.onDomainChanged with the selected range", () => {
        instance._getD3Event = () => ({
            sourceEvent: {
                type: "brush"
            },
            selection: [0, 100]
        });

        instance._onResizeBrush();

        expect(onDomainChanged.mock.calls.length).toBe(3);
        expect(onDomainChanged.mock.calls[2][0]).toEqual([0, 100]);
    });

    it("calls props.onDomainChanged with the whole range if the selection was empty", () => {
        instance._getD3Event = () => ({
            sourceEvent: {
                type: "brush"
            }
        });

        instance._onResizeBrush();

        expect(onDomainChanged.mock.calls.length).toBe(3);
        expect(onDomainChanged.mock.calls[2][0]).toEqual([0, 1000]);
    });
});

describe("componentDidUpdate", () => {
    it("update the brush when the screen have been resized", () => {
        const wrapper = mount(<DensityChart
            data={smallSample}
            width={width}
            height={50}
            padding={10}
            brushDomainMax={brushDomainMax}
            brushDomainMin={brushDomainMin}
            densityChartXScale={densityChartXScale}
            onDomainChanged={onDomainChanged}
            xAccessor={xAccessor}
            renderPlayButton={false}
        />);
        const brushExtendMock = jest.fn();
        const instance = wrapper.instance();
        const newWidth = width - 50;

        instance.brush.extent = brushExtendMock;

        wrapper.setProps({ width: newWidth });

        expect(brushExtendMock).toHaveBeenCalledTimes(1);
        expect(brushExtendMock).toHaveBeenCalledWith([
            [0, 0],
            [newWidth, 50]
        ]);
    });
});

describe("render", () => {
    it("does a baseline render", () => {
        expect(mount(<DensityChart
            data={smallSample}
            width={width}
            height={50}
            padding={10}
            brushDomainMax={brushDomainMax}
            brushDomainMin={brushDomainMin}
            densityChartXScale={densityChartXScale}
            onDomainChanged={onDomainChanged}
            xAccessor={xAccessor}
        />)).toMatchSnapshot();
    });

    it("doesn't render the play button if renderPlayButton is false", () => {
        expect(mount(<DensityChart
            data={smallSample}
            width={width}
            height={50}
            padding={10}
            brushDomainMax={brushDomainMax}
            brushDomainMin={brushDomainMin}
            densityChartXScale={densityChartXScale}
            onDomainChanged={onDomainChanged}
            xAccessor={xAccessor}
            renderPlayButton={false}
        />).find(PlayButton).length).toBe(0);
    });
});

afterEach(() => {
    global.d3Event = d3Event;
});
