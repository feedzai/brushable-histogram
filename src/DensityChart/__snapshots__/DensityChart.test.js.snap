// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render does a baseline render 1`] = `
<DensityChart
  brushDensityChartColor="rgba(33, 150, 243, 0.2)"
  brushDensityChartFadedColor="rgba(176, 190, 197, 0.2)"
  brushDomainMax={1534164400000}
  brushDomainMin={1533309900034}
  data={
    Array [
      Object {
        "timestamp": 1534164400000,
        "total": 0,
      },
      Object {
        "timestamp": 1533704702305,
        "total": 0.8414709848078965,
      },
      Object {
        "timestamp": 1533580546837,
        "total": 0.9092974268256817,
      },
      Object {
        "timestamp": 1534154392497,
        "total": 0.1411200080598672,
      },
      Object {
        "timestamp": 1533818043621,
        "total": 0.7568024953079282,
      },
      Object {
        "timestamp": 1533448062185,
        "total": 0.9589242746631385,
      },
      Object {
        "timestamp": 1534124570286,
        "total": 0.27941549819892586,
      },
      Object {
        "timestamp": 1533918302254,
        "total": 0.6569865987187891,
      },
      Object {
        "timestamp": 1533309900034,
        "total": 0.9893582466233818,
      },
      Object {
        "timestamp": 1534075530262,
        "total": 0.4121184852417566,
      },
    ]
  }
  densityChartXScale={[Function]}
  height={50}
  onDomainChanged={
    [MockFunction] {
      "calls": Array [
        Array [
          Array [
            0,
            1000,
          ],
        ],
        Array [
          Array [
            0,
            1000,
          ],
        ],
      ],
      "results": Array [
        Object {
          "isThrow": false,
          "value": undefined,
        },
        Object {
          "isThrow": false,
          "value": undefined,
        },
      ],
    }
  }
  overallTimeDomainMax={-Infinity}
  padding={10}
  renderPlayButton={true}
  width={1000}
  xAccessor={[Function]}
>
  <div
    className="fdz-css-graph-histogram-density__wrapper"
  >
    <PlayButton
      brushDomainMax={1534164400000}
      brushDomainMin={1533309900034}
      densityChartXScale={[Function]}
      frameDelay={500}
      frameStep={0.025}
      moveBrush={[Function]}
      width={1000}
    >
      <button
        className="ant-btn fdz-css-play-btn ant-btn-icon-only"
        onClick={[Function]}
        type="button"
      >
        <i
          className="anticon anticon-play-circle"
        >
          <svg
            fill="currentColor"
            height="1em"
            viewBox="64 64 896 896"
            width="1em"
          >
            <path
              d="M512 64C264.6 64 64 264.6 64 512s200.6 448 448 448 448-200.6 448-448S759.4 64 512 64zm0 820c-205.4 
    0-372-166.6-372-372s166.6-372 372-372 372 166.6 372 372-166.6 372-372 372z"
            />
            <path
              d="M719.4 499.1l-296.1-215A15.9 15.9 0 0 0 398 297v430c0 13.1 14.8 20.5 25.3 12.9l296.1-215a15.9 
    15.9 0 0 0 0-25.8zm-257.6 134V390.9L628.5 512 461.8 633.1z"
            />
          </svg>
        </i>
      </button>
    </PlayButton>
    <div
      className="fdz-css-graph-histogram-density"
      style={
        Object {
          "position": "relative",
        }
      }
    >
      <canvas
        aria-label="Density Chart"
        className="fdz-css-graph-histogram-density__canvas"
        height={50}
        style={
          Object {
            "left": 0,
          }
        }
        width={1000}
      />
      <svg
        alt="Density Chart Brush"
        className="fdz-css-graph-histogram-brush"
        height={50}
        style={
          Object {
            "left": 0,
            "position": "absolute",
            "top": 0,
          }
        }
        width={1000}
      />
    </div>
  </div>
</DensityChart>
`;
